Critter combined tree

Jacob-Rueckert opened the pull request:
I combined several critter behaviors into one. 

Based on the student project in cooperation with @jdrueckert  @skaldarnar 
- made a priority list of the current behaviors
- added features from MetalRenegadesCitizen

Goal is to have an universal behavior which differs based on their configuration not their implementation.

Jacob-Rueckert committed:
doc: add concept trees for AI

source files for tree visualisations in issue:https://github.com/Terasology/Behaviors/issues/110

Jacob-Rueckert committed:
doc: replace empty files

Jacob-Rueckert committed:
doc(ai_concepts): add prioritised combination of subtrees

add 4 Subtrees into megatree with sequential prioritizaition

Jacob-Rueckert committed:
feat(behavior): add combination of critter behavior trees

skaldarnar commented on assets/behaviors/creatures/combinedCritter.behavior:
```diff
@@ -0,0 +1,107 @@
+  // First Guard fo Selector for Territory End.
+    guard: { // Implementation of scared Critter
+        componentPresent: "Behaviors:FindNearbyPlayers",
+        values: ["N charactersWithinRange != null"], // Adjust for Nemesis / Scary Creatures
```
We discussed this in the meeting today.

The `N` is a parsing directive that specifiies whether the condition has the form `type fieldName operator` or `type fieldName operator valueOrReference`.

Technically, we could use `V charactersWithinRange != null` to ensure that the list itself is defined and not `null` (checking agains the _value_ `null`). However, this is not the semantics we are looking for.

Therefore, using the `nonEmpty` operator on the collection with no value to check against (`N`) is the solution:
```suggestion
        values: ["N charactersWithinRange nonEmpty"], // Adjust for Nemesis / Scary Creatures
```

skaldarnar commented on assets/behaviors/creatures/combinedCritter.behavior:
```diff
@@ -0,0 +1,107 @@
+      // Territory would to be implemented as Component
+      // Implementation of territorial Critter
+      componentPresent: "Behaviours:Territory",
+      values: ["F Territory != null"],
```
This would try to make a lookup of a _field_ (indicated by `F`) named `null`.

However, I believe you want to check against the _value_ `null`, thus using `V` would be correct.

```suggestion
      values: ["V Territory != null"],
```

Alternatively, you can use the `exists` operator (basically `... != null`) without an explicit value to check against:

```suggestion
      values: ["N Territory exists"],
```

skaldarnar replied:
With the comment above about the component, this should most likely be 
```suggestion
      values: ["F location exists"],
```

skaldarnar commented on assets/behaviors/creatures/combinedCritter.behavior:
```diff
@@ -0,0 +1,107 @@
+
+      // Territory would to be implemented as Component
+      // Implementation of territorial Critter
+      componentPresent: "Behaviours:Territory",
```
I don't think the `TerritoryComponent` exists in the Behaviors module.

The closest thing we have is [TerritoryDistance](https://github.com/Terasology/Behaviors/blob/develop/src/main/java/org/terasology/module/behaviors/components/TerritoryDistance.java), a component with two fields `distance` and `location`.

Therefore, the referenced component should probably be 
```suggestion
      componentPresent: "Behaviours:TerritoryDistance",
```

and the referenced field below cannot be `Territory` (as it does not exist) but either `distance` or `location`. I suspect that we can check whether the location is set.

Jacob-Rueckert committed:
feat: add LureComponent, delete constant in CheckLuringItemInUseAction

change behaviour system from hard coded to configurable

Jacob-Rueckert committed:
feat: add FindNearbyNemesisComponent & FleeFromNemesisAction

Use for configuration of different entities to flee from
Implement Action in combinedCritter.behavior for new component

Jacob-Rueckert committed:
feat: detail out territorial behavior guards

adressing review comments

Jacob-Rueckert committed:
fix: missing import and JSon Format

Jacob-Rueckert committed:
fix: resolve 2 syntax errors

game crash on BehaviourTree initialisation because of NullPointerException

Jacob-Rueckert committed:
refactor: add TerritoryComponent, revise TerritorySystem, add component to sheepBase

combinedCritter.behaviour crash on Territory stage cause of missing parameters

Jacob-Rueckert commented on src/main/java/org/terasology/module/behaviors/systems/TerritorialBehaviourSystem.java:
```diff
@@ -45,18 +50,40 @@ public void initialise() {
+        creature.saveComponent(territoryComponent);
+    }
+
+    /*
```
Is this unnecessary when we look at the update Function above?

Jacob-Rueckert commented on src/main/java/org/terasology/module/behaviors/systems/TerritorialBehaviourSystem.java:
```diff
@@ -45,18 +50,40 @@ public void initialise() {
-                entity.getComponent(LocationComponent.class).getWorldPosition(new Vector3f()));
-            entity.saveComponent(territoryDistance);
+        for (EntityRef entity : entityManager.getEntitiesWith(TerritoryComponent.class, LocationComponent.class)) {
+
```
So is this how it its supposed to work for DistanceSquared?

jdrueckert committed:
Merge remote-tracking branch 'origin/develop' into AI_Combined_Tree

Jacob-Rueckert committed:
chore: safe for Progress

Jacob-Rueckert committed:
fix: try behavior instead of Behaviours

Jacob-Rueckert committed:
Merge remote-tracking branch 'origin/develop' into AI_Combined_Tree

Jacob-Rueckert committed:
refactor: syntax error changes

NullPointerExceptions caused by missspelling

Jacob-Rueckert committed:
fix: use naiveMoveTo instead of moveTo Action

moveTo sets GoalEntity but doesnt update GoalPosition and path
naiveMoveTo does

probably originates from pathfinding migration

Jacob-Rueckert committed:
fix: behavior errors

using field instead of component

Jacob-Rueckert committed:
chore: initilize distanceSquared with 0
